### PR TITLE
[language/en] Add ada

### PR DESCRIPTION
**Pull request by Lucretia**
- [X] I solemnly swear that this is all original content of which I am the original author
- [X] Pull request title is prepended with `[language/lang-code]` (example `[python/fr-fr]` or `[java/en]`)
- [X] Pull request touches only one file (or a set of logically related files with similar changes made)
- [X] Content changes are aimed at *intermediate to experienced programmers* (this is a poor format for explaining fundamental programming concepts)
- [X] If you've changed any part of the YAML Frontmatter, make sure it is formatted according to [CONTRIBUTING.md](https://github.com/adambard/learnxinyminutes-docs/blob/master/CONTRIBUTING.markdown)
  - [X] Yes, I have double-checked quotes and field names!

### CODE REVIEW

File: ada.html.markdown [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+
+   --  You can force conversions, but then you are warned by the name of the
+   --  package that you may be doing something unsafe.
+   function Degrees_To_Int is new Ada.Unchecked_Conversion
+     (Source => Degrees,   --  Line continuations are indented by 2 spaces.
+      Target => Integer);
+
+   New_Value_2 : Integer := Degrees_To_Int (Angle);  --  Note, space before (.
+
+   --  Ada has a style guide and GNAT will force you to adhere to it, so that
```
Comment (Fabien-Chouteau) [inline ada.html.markdown:161]: This is missing a definition of what GNAT is.
Also GNAT doesn't force a style, but it provides options to check coding style.

File: ada.html.markdown [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+
+   --  You can define the array's range using the 'Range attribute from
+   --  another type.
+   type Vector is array (Axes'Range) of Float;
+
+   V1 : constant Vector := (0.0, 0.0, 1.0);
+
+   --  A record is the same as a structure in C, C++.
+   type Entities is record
+      Name     : String (1 .. 10);  --  Always start at 1, inclusive range.
```
Comment (Fabien-Chouteau) [inline ada.html.markdown:207]: String doesn't always start at 1. One can define `String (2 .. 11)`.

File: ada.html.markdown [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+      Position : Vector;
+   end record;
+
+   --  In Ada, you have to pad out the full string. Change the number of spaces
```
Comment (Fabien-Chouteau) [inline ada.html.markdown:211]: I'd like to suggest a different explanation here:
```ada
   --  In Ada, array bounds are immutable. You therefore have to provide a
   --  string literal with a value for every character.
   E1 : constant Entities := ("Blob ", (0.0, 0.0, 0.0));

   --  An alternative is to use an array aggregate and assign a default value to
   --  every element that wasn't previously assigned in this aggregate.
   --  ```others``` is used to indicate anything else that has not been explicitly initialized.
   E2 : constant Entities := (('B', 'l', 'o', 'b', others => ' '), (0.0, 0.0, 0.0));

   -- There are [dynamic length strings](https://ada-lang.io/docs/arm/AA-A/AA-A.4#Subclause_A.4.5) available in the standard library.
```

File: ada.html.markdown [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+$ gnatchop learn-ada-in-y.ada # This breaks the program into its specificication ".ads" and body ".adb".
+$ gnatmake empty.adb # gnatmake takes care of compilation of all units and linking.
+$ gnatmake hello.adb
+$ gnatmake learnadainy.adb
+```
+
+Or, download [Alire](https://alire.ada.dev), copy it to somewhere in your PATH and then do this:
+
+```bash
+$ alr toolchain --select    # Select the GCC and GPR tools to install
```
Comment (Fabien-Chouteau) [inline ada.html.markdown:345]: This not necessary as Alire 1 will ask the user anyway, and Alire 2 will use defaults.

File: ada.html.markdown [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+$ alr run hello
+$ alr run learnadainy
+```
+
+## Further Reading
+
+* [Ada Language](https://ada-lang.io)
+* [Ada 2022 Reference Manual](https://ada-lang.io/docs/arm)
+* [Ada Style Guide](https://ada-lang.io/docs/style-guide/Ada_Style_Guide)
+
```
Comment (Fabien-Chouteau) [inline ada.html.markdown:359]: I would add https://learn.adacore.com/ as it is the logical next step after this introduction.

**Review by Fabien-Chouteau — COMMENTED** [review]
Great job @Lucretia and @Irvise ! It's nice to see Ada here.

File: ada.html.markdown (hunk 1 above)
Comment (Lucretia) [reply ada.html.markdown:161]: Ok, maybe nag you to fix it.

File: ada.html.markdown (hunk 5 above)
Comment (Lucretia) [reply ada.html.markdown:387]: All added, thanks.

File: ada.html.markdown [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+   Do_Something;
+end Stuff;
+
+
+with Ada.Unchecked_Conversion;
+with Ada.Text_IO;
+with Stuff;
+
+procedure LearnAdaInY is
+   --  Indentation is 3 spaces.
```
Comment (mgrojo) [inline ada.html.markdown:118]: Typical indentation is 3 spaces
Conventional? Common?

**Review by mgrojo — COMMENTED** [review]
Thanks for doing this. I leave you some comments and a spelling review, so it is in the best shape.

File: ada.html.markdown (hunk 6 above)
Comment (Lucretia) [reply ada.html.markdown:118]: Convention. Style Guide.

File: ada.html.markdown [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+   Blue_Hue   :          Primaries := Blue;  --  A variable.
+   Red_Hue    : constant Primaries := Red;   --  A constant.
+   Yellow_Hue : constant Hues      := Yellow;
+   Colour_1   : constant Hues      := Red_Hue;
+   --  Colour_2   : constant Primaries := Yellow_Hue;  --  uncomment to compile.
+
+   --  You can force conversions, but then you are warned by the name of the
+   --  package that you may be doing something unsafe.
+   function Degrees_To_Int is new Ada.Unchecked_Conversion
+     (Source => Degrees,   --  Line continuations are indented by 2 spaces.
```
[the remark this answers is not in the document]
Comment (Lucretia) [reply ada.html.markdown:171]: Yes.

**Comment by Lucretia** [comment]
Can this be merged in now?

**Review by vendethiel — COMMENTED** [review]
Small nitpicks.

File: ada.html.markdown [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+   --  Strings are arrays. Note how parentheses are used to access elements of
+   --  an array? This is mathematical notation and was used because square
+   --  brackets were not available on all keyboards at the time Ada was
+   --  created. Also, because an array can be seen as a function from a
+   --  mathematical perspective, so it made converting between arrays and
+   --  functions easier.
+   Char : constant Character := Str (Str'First);  --  ```'First``` is a type
+                                                  --  attribute.
+
+   --  Ada 2022 includes the use of [] for array initialisation when using
```
[the remark this answers is not in the document]
Comment (Lucretia) [reply ada.html.markdown:211]: No, it's a really new addition, literally within the last few months the spec has been finalised. It follows on from () as array indexing.

File: ada.html.markdown [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+
+   --  If there are sub-programs declared within the specification, the body
+   --  of the sub-program must be declared within the package body.
+   procedure Do_Something;  --  If a subprogram takes no parameters, empty
+                            --  parentheses are not required, unlike other
+                            --  languages.
+
+   --  We can also make generic sub-programs.
+   generic
+      type Element is (<>);
```
[the remark this answers is not in the document]
Comment (Lucretia) [reply ada.html.markdown:72]: It's different notation for a different thing.

**Comment by vendethiel** [comment]
Amazing, thank you very much. Looks good on my end, tell me when I should merge this !

**Comment by Lucretia** [comment]
> Amazing, thank you very much. Looks good on my end, tell me when I should merge this !

Do it now if you want.

**Comment by Lucretia** [comment]
> Thankyou again!

Thanks.

How long does it take to get pushed to the site?

**Comment by vendethiel** [comment]
It pulls every so often, unsure how long until the next refresh

**Comment by mgrojo** [comment]
Still not there?

**Pull request merged**